manual/manual/tutorials/polymorphism.etex

```diff
@@ -0,0 +1,387 @@
+let id_again = fun x -> (fun x -> x) (fun x -> x) x;;
+\end{caml_example}
+With this argument, "id_again" is seen as a function definition by the type
+checker and can therefore be generalized.
+
+This kind of manipulation is called eta-expansion in lambda calculus and
+is sometimes referred under this name.
+
+\subsection{The relaxed value restriction}
+
+There is another partial solution to the problem of unnecessary weak type
+in the type of value, which is implemented directly within the type
+checker. Briefly, it is possible to prove that weak types that only appear
+as type parameters in a covariant position can be safely generalized to
+polymorphic types. For instance, the type "'a  list" is covariant in "'a":
+\begin{caml_example}
+  let f () = [];;
+  let empty = f ();;
+\end{caml_example}
+Remark that the type inferred for "empty" is "'a list" and not "'_a list"
+that should have occurred with the value restriction since "f ()" is a
+function application.
+
+The value restriction combined with this generalization for covariant type
+parameters is called the relaxed value restriction.
+
+%todo: once the manual contains a detailed covariant/contravariant
+% subsection, add a link to this part
+\subsection{Covariance and value restriction}
+Remember that a type constructor "'a t", for instance
+\begin{caml_example}
+  type 'a tree = Leaf of 'a | Branch of 'a tree * 'a tree;;
+\end{caml_example}
+is covariant in its parameter "'a" if it preserves the order of the
+subtyping relationship. As an illustration, we can construct a pair of type
+"x" and "xy" with "xy" a subtype of "x", i.e "xy :> x":
+\begin{caml_example}
+  type x = [ `X ];;
+  let x:x = `X;;
+  type xy = [ `X | ` Y ];;
+  let x' = ( x :> xy);;
+\end{caml_example}
+Then, if the type constructor "'a t" is covariant in "'a",
+the subtyping relationship "x :> xy" can be lifted to "x t :> xy t".
+
+This notion of covariance can be used to prove that a weak type"'_a" that
+only appears in covariant position in a type expression can be safely
+generalized to a polymorphic type.
+
+Let's denote this type expression by "'_a t" as an abuse of notation. Then,
+since the type "'_a" is not yet known, we can replace it by any type of
+our convenience. In particular, we can replace "'_a" by "empty" where "empty"
+is a type with no existing values. Since "empty" appears only in covariant
+position inside "empty t", we can then use covariance to replace "empty" by
+any of its supertypes. A property of the empty set is that all its elements
+(i.e. none) belongs to any type, all types are therefore a supertype of
+"empty". Consequently, we can use covariance to replace "empty" by a
+polymorphic type "'a" and obtain back a polymorphic type "'a t".
+
+Allied with this relaxed value restriction, type parameter covariance
+helps to avoid eta-expansion in many situation.
+
+\subsection{Abstract data types}
+Moreover, when the associated type definitions are visible, the type checker
```

Comment:
Let's avoid "associated type" which means something else. You could just say "When the type definitions are exposed".